rasterio/mask.py

```diff
@@ -187,9 +192,12 @@ def mask(dataset, shapes, all_touched=False, invert=False, nodata=None,
         out_shape = (len(indexes), ) + shape_mask.shape
 
     out_image = dataset.read(
-        window=window, out_shape=out_shape, masked=True, indexes=indexes)
+        window=window, out_shape=out_shape, masked=masked, indexes=indexes)
 
-    out_image.mask = out_image.mask | shape_mask
+    if masked:
+        out_image.mask = out_image.mask | shape_mask 
+    else:
+        out_image = numpy.ma.masked_array(out_image, mask = shape_mask)
```

Comment:
@AntoineGuiot @snowman2 if I read these right, mask always returns a masked array even if masked=False. That's not consistent with the docstring.